File: android/src/main/java/sh/autograph/Cert.kt
package sh.autograph

internal class Cert {
    companion object {
        init {
            System.loadLibrary("autograph")
        }

        private external fun autographSignatureSize(): Int

        private external fun autographCertify(
            signature: ByteArray,
            ourIdentityKeyPair: ByteArray,
            theirIdentityKey: ByteArray,
            data: ByteArray,
        ): Boolean

        private external fun autographVerify(
            ownerIdentityKey: ByteArray,
            certifierIdentityKey: ByteArray,
            signature: ByteArray,
            data: ByteArray,
        ): Boolean

        fun createSignature(): ByteArray = ByteArray(autographSignatureSize())

        fun certify(
            ourIdentityKeyPair: ByteArray,
            theirIdentityKey: ByteArray,
            data: ByteArray?,
        ): ByteArray {
            val signature = createSignature()
            val success =
                autographCertify(
                    signature,
                    ourIdentityKeyPair,
                    theirIdentityKey,
                    data ?: byteArrayOf(),
                )
            if (!success) {
                throw RuntimeException("Certification failed")
            }
            return signature
        }

        fun verify(
            ownerIdentityKey: ByteArray,
            certifierIdentityKey: ByteArray,
            signature: ByteArray,
            data: ByteArray?,
        ): Boolean {
            return autographVerify(
                ownerIdentityKey,
                certifierIdentityKey,
                signature,
                data ?: byteArrayOf(),
            )
        }
    }
}

public fun certify(
    ourIdentityKeyPair: ByteArray,
    theirIdentityKey: ByteArray,
    data: ByteArray?,
): ByteArray {
    return Cert.certify(ourIdentityKeyPair, theirIdentityKey, data)
}

public fun verify(
    ownerIdentityKey: ByteArray,
    certifierIdentityKey: ByteArray,
    signature: ByteArray,
    data: ByteArray?,
): Boolean {
    return Cert.verify(ownerIdentityKey, certifierIdentityKey, signature, data)
}

File: apple/Sources/Autograph/Cert.swift
import Clibautograph
import Foundation

func createSignature() -> [UInt8] {
    createBytes(autograph_signature_size())
}

public func certify(
    _ ourIdentityKeyPair: [UInt8],
    _ theirIdentityKey: [UInt8],
    _ data: [UInt8]?
) throws -> [UInt8] {
    var signature = createSignature()
    var success = false
    if let data {
        success = autograph_certify(
            &signature,
            ourIdentityKeyPair,
            theirIdentityKey,
            data,
            data.count
        )
    } else {
        success = autograph_certify(
            &signature,
            ourIdentityKeyPair,
            theirIdentityKey,
            createBytes(0),
            0
        )
    }
    if !success {
        throw AutographError.certification
    }
    return signature
}

public func verify(
    _ ownerIdentityKey: [UInt8],
    _ certifierIdentityKey: [UInt8],
    _ signature: [UInt8],
    _ data: [UInt8]?
) -> Bool {
    var verified = false
    if let data {
        verified = autograph_verify(
            ownerIdentityKey,
            certifierIdentityKey,
            signature,
            data,
            data.count
        )
    } else {
        verified = autograph_verify(
            ownerIdentityKey,
            certifierIdentityKey,
            signature,
            createBytes(0),
            0
        )
    }
    return verified
}
